### PR TITLE
builds, pushes kfctl to GCS as part of pre and post submits

### PR DESCRIPTION
**Pull request by swiftdiaries**
Things investigated:

- [x] TektonCD to build kfctl
- [x] Kaniko to build kfctl 
- [x] Argo Cron Jobs
- [x] Prow periodic jobs

TektonCD and Kaniko failed with similar errors while building kfctl. The way the files are mapped either from git checkout to build step in Tekton and local filesystem to container volume for kaniko is buggy (or likely I made a mistake). 
go.mod file is not found in the context in Tekton and the hack/v2.zip is not found in the kaniko build. 
Argo Cron Jobs need to be run as Kubernetes Cron Jobs with the workflow YAML pulled in and `kubectl create`'d on the fly. In our current workflow, running that would be to do a `ks show <app_name> -c <workflow_name> | kubectl create -f -`.
Prow periodic jobs are run like [this 
](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubeflow/kubeflow-periodics.yaml#L3) but I wasn't sure how to add this field from the run_e2e_workflow.py in kubeflow/testing.

I ran these on latest master.

In the interim, I've added an extra workflow to build and push to GCS as part of pre-submits.
I'll continue to work towards making it a nightly build and push. 

Any ideas / thoughts would be good. Please add to discussion here: https://github.com/kubeflow/kubeflow/issues/3662.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/3949)

### CODE REVIEW

**Comment by jlewi** [comment]
Why presubmits why not postsubmits?

**Comment by jlewi** [comment]
Maybe we should do it on pre and postsubmit builds just to ensure the logic gets tested on presubmits?

**Comment by swiftdiaries** [comment]
We can do both pre and post submits

File: prow_config.yaml [hunk 1]
```diff
@@ -63,6 +63,25 @@ workflows:
       workflowName: kfctl-go
       useBasicAuth: false
       useIstio: true
+  # Run a nightly build of kfctl and push the binary to GCS    
+  - app_dir: kubeflow/kubeflow/testing/workflows
```
Comment (jlewi) [inline prow_config.yaml]: Why do we need a new prow job? It looks like we are copying the binary in kfctl_go_test.py.
That gets executed on every pre/post and periodic run.
Isn't that sufficient?

File: testing/kfctl/kfctl_go_test.py [hunk 2]
```diff
@@ -65,6 +65,18 @@ def test_build_kfctl_go(app_path, project, use_basic_auth, use_istio, config_pat
   run_with_retries(["make", "build-kfctl"], cwd=build_dir)
   kfctl_path = os.path.join(build_dir, "bin", "kfctl")
 
+  # Push built kfctl binary to GCS bucket
+  import google.cloud.storage as storage
```
Comment (jlewi) [inline testing/kfctl/kfctl_go_test.py]: Why is this import here and not at the top of the file?

File: testing/kfctl/kfctl_go_test.py [hunk 3]
```diff
@@ -65,6 +65,18 @@ def test_build_kfctl_go(app_path, project, use_basic_auth, use_istio, config_pat
...
+  gcs_client = storage.Client()
+  bucket = gcs_client.lookup_bucket("kfctl-nightly")
```
Comment (jlewi) [inline testing/kfctl/kfctl_go_test.py]: Should we just use a GCS bucket and then do different subpaths?
We could also just copy kfctl to the bucket that is used by prow.
Then the binaries would show up in the artifacts tab of spyglass e.g.
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubeflow_kubeflow/3856/kubeflow-presubmit/1165454527155933184/

And people could get to them just by navigating to the appropriate pre/post/periodic  run.

**Comment by jlewi** [comment]
@swiftdiaries It looks like the commit history is messed up. It looks like you may have done a merge and pulled in changes that are not your own. Can you rebase on master and drop the merge commits?

**Comment by jlewi** [comment]
@swiftdiaries Any update on this? Would be great to wrap it up.

**Comment by jlewi** [comment]
@swiftdiaries ping?
